hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionReplicaStateChecker.java

```diff
@@ -0,0 +1,386 @@
+        if (masterAddress == null) {
+            return node.joined();
+        }
+        Operation operation = new HasOngoingMigration();
+        Future future = invokeOnTarget(operation, masterAddress);
+        try {
+            return (Boolean) future.get(1, TimeUnit.MINUTES);
+        } catch (InterruptedException ie) {
+            Logger.getLogger(InternalPartitionServiceImpl.class).finest("Future wait interrupted", ie);
+        } catch (Exception e) {
+            logger.log(level, "Could not get a response from master about migrations! -> " + e.toString());
+        }
+        return false;
+    }
+
+    private Future invokeOnTarget(Operation operation, Address target) {
+        OperationService operationService = nodeEngine.getOperationService();
+        return operationService.createInvocationBuilder(SERVICE_NAME, operation, target)
+                .setTryCount(INVOCATION_TRY_COUNT)
+                .setTryPauseMillis(INVOCATION_TRY_PAUSE_MILLIS)
+                .invoke();
+    }
+
+    private boolean isReplicaInSyncState() {
+        if (!partitionStateManager.isInitialized() || partitionStateManager.getMemberGroupsSize() == 0) {
+            return true;
+        }
+
+        final int replicaIndex = 1;
+        final List<Future> futures = new ArrayList<Future>();
+        final Address thisAddress = node.getThisAddress();
+        for (InternalPartition partition : partitionStateManager.getPartitions()) {
+            final Address owner = partition.getOwnerOrNull();
+            if (owner == null) {
+                return false;
+            } else if (thisAddress.equals(owner)) {
+                if (partition.getReplicaAddress(replicaIndex) != null) {
+                    int partitionId = partition.getPartitionId();
+                    long replicaVersion = getCurrentReplicaVersion(replicaIndex, partitionId);
+                    Operation operation = createReplicaSyncStateOperation(replicaVersion, partitionId);
+                    Future future = invoke(operation, replicaIndex, partitionId);
+                    futures.add(future);
+                }
+            }
+        }
+
+        for (Future future : futures) {
+            boolean isSync = getFutureResult(future, REPLICA_SYNC_CHECK_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+            if (!isSync) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private Future invoke(Operation operation, int replicaIndex, int partitionId) {
+        OperationService operationService = nodeEngine.getOperationService();
+        return operationService.createInvocationBuilder(SERVICE_NAME, operation, partitionId)
+                .setTryCount(INVOCATION_TRY_COUNT)
+                .setTryPauseMillis(INVOCATION_TRY_PAUSE_MILLIS)
+                .setReplicaIndex(replicaIndex)
+                .invoke();
+    }
+
+    // TODO: VISIBILITY PROBLEM! Replica versions are updated & read only by partition threads!
```

Comment:
This will be fixed by graceful shutdown work. Currently, this part reads replica versions from a non-partition thread without guaranteeing thread safety. Normally they are accessed only within partition threads.